### ISSUE TITLE
[Feature Request] Email Integration

### ISSUE THREAD

**Issue by scotwells** [datum-cloud/enhancements#80]
## Summary

Datum OS needs the ability to send emails to end-users based on events happening across the Datum OS platform. 

### Potential integrations to evaluate

- https://resend.com/
- https://loops.so/

## Goals

- Maintain control over source of truth information (e.g. marketing contacts)
- Create a system that would easily support integrations with new email providers in the future without having to change core functions of the Datum OS platform
- Create a feedback loop that allows us to have a single pane of glass view on emails being sent to contacts by all configured email providers

## Non-Goals

- Re-implement functionality that doesn't align with the core of Datum OS or Datum Cloud (e.g. managing email templates or designs)
  - It may be desirable to configure the content that will be sent with a given email template, but that should be done through template variables and not by managing the template directly. Templates should always be managed in the email provider.

## User Stories

Below are a list of user stories that could be implemented by the completion of this integration. 

- As a Datum OS staff user, I should be able to view all of the email activity for any user that's registered with the platform
- As a Datum OS staff user, I should be able to view the history of an email along with it's current status (e.g. Sent, Opened, Soft-Bounced, Hard-Bounced, Spam, Unsubscribed, etc)
- As a Datum OS staff user, I should be able to configure Datum OS platform events to be forwarded to the configured email provider to help orchestrate marketing or engagement campaigns I've setup. 
  - e.g. I create a [new Loop](https://loops.so/docs/loop-builder) that will send users an email when customer creates a new payment method. I can configure an event (e.g. `datum.os.billing.CreatePaymentMethod`) to be sent to Loops (see [sending events](https://loops.so/docs/events) so Loops knows to send the email.

## Considerations

- Much of this document is scoped to sending emails only by reacting to events but we should consider a future where additional types of notifications could be sent to a user (e.g. SMS, Webhook, Pub/Sub, etc.)
- Emails will be sent to marketing contacts (subscribed to a contact list, but potentially not a platform user) as well as registered users of the platform. We should ensure we're correctly handling all types of emails.

## Dependencies

- **Events**: Datum OS needs a generic eventing platform we can hook into so we can forward events to configured email providers (e.g. Loops) to trigger emails / marketing campaigns.
- **Workflows**: We need a durable workflow engine we can use to react to events, enrich with data, and forward events to external systems like Loops.

### Potential eventing architecture

> [!NOTE]
>
> The sequence diagram below attempts to highlight the separation of responsibilities we want to maintain in the architecture to keep the integration flexible and make it possible to support new email providers in the future. This does **NOT** suggest the decided architecture which is expected to evolve as additional requirements are learned.

```mermaid
sequenceDiagram
    actor user as Datum Cloud User
    participant datumCloudPortal as Datum Cloud
    participant datumAPI as Datum API
    participant emailProviderHandler as Email Provider Event Handler
    participant events as Event Bus
    participant workflow as Workflow Engine
    participant emailProvider as Email Provider (e.g. Loops)

    user ->> datumCloudPortal: Add payment method to billing account
    datumCloudPortal ->> datumAPI:  Add Payment Method
    datumAPI ->> events: Payment Method Added

    events -->> workflow: Receives Payment Method added event
    workflow ->> emailProvider: Send Payment Method added event

    emailProvider ->> user: Sends payment method added email notification
    emailProvider ->> emailProviderHandler: Send email_sent event to platform
    emailProviderHandler ->> datumAPI: Update email activity log

    user -->> emailProvider: Opens Email

    emailProvider -->> emailProviderHandler: Sends email_opened event
    emailProviderHandler ->> datumAPI: Updates email acticity log
```

A new email provider would require the following development work:
- Creating a new **Email Provider Event Handler** that knows how to receive events from webhooks sent by the email provider and update Datum OS using standardized APIs.
- Adding new workflow definitions to listen for events happening on Datum OS and forward the event to the email provider with the appropriate context necessary for the provider to send the email.
  - Workflows would also need to be added to sync contact information over to the email provider based on changes being made to Datum OS. (e.g. new contact is added because of user registration) 
- Setting up the email provider to send emails based on events received from Datum OS.

**Comment by zsmithnyc** [comment]
I want to reinforce the concept that communication sent via email is one of the potential avenues and our architecture for events should allow for this. Push notification (mobile or web), in app, webhook, etc should be considered as whether we do should easily support these methods and not force us to reimplement significant functionality.

**Comment by zsmithnyc** [comment]
Yes, that makes sense and what I'm attempting to describe.  So do we have a notification underlay here that we'll be building this email transaction stuff off of?

**Comment by JoseSzycho** [comment]
@scotwells Regarding the [ Contacts](https://resend.com/docs/api-reference/contacts/create-contact),[Audiences](https://resend.com/docs/api-reference/audiences/create-audience) and [Broadcast](https://resend.com/docs/api-reference/broadcasts/create-broadcast), do we want to let the email provider handle sending broadcast emails, or is this type of large-scale email sending something we’d prefer to manage ourselves using their basic [Send Email](https://resend.com/docs/api-reference/emails/send-email) or [Send Batch Emails](https://resend.com/docs/api-reference/emails/send-batch-emails)?

I think we should create the Audience, Contact, and Broadcast CRDs, and implement controllers that synchronize these CRs with the email provider. So the email provider would be in charge of Broadcasts.

But want to hear your opinion on that

**Comment by scotwells** [comment]
@JoseSzycho we would introduce new concepts in Milo that represent contacts, contact groups, and announcements sent out to contact groups.

For contact groups, I was thinking it would be good if there was a Contact Group Policy or some other resource that a user can configure to automatically manage the list of contacts that make up a contact group. This way we have automated management of contact groups for things like billing account contacts, platform users, etc while maintaining the ability to manually build a contact list.

**Comment by JoseSzycho** [comment]
@scotwells that's sound greats, now I have a clear idea on how to proceed

**Comment by JoseSzycho** [comment]
## Quick update

---

Main tasks have been completed. 

- [Define Email CRD in Milo #239](https://github.com/datum-cloud/enhancements/issues/239)

- [Define Email Template CRD in Milo #240](https://github.com/datum-cloud/enhancements/issues/240)

- [Create Validators and Mutator webhook for Email CRs in Milo #241](https://github.com/datum-cloud/enhancements/issues/241)

- [Create Validator webhook for EmailTemplate in Milo #242](https://github.com/datum-cloud/enhancements/issues/242)

- [Create controller for sending emails in email-provider-resend #243](https://github.com/datum-cloud/enhancements/issues/243)

- [Create EmailEvent CRD in Milo #244](https://github.com/datum-cloud/enhancements/issues/244)

- [Create server for receiving Resend webhooks in email-provider-resend #245](https://github.com/datum-cloud/enhancements/issues/245)

Creating two new sub issues, a deployment one, and other for tackling unit tests for the `email-controller` and the `resend-webhook`
